gobblin-runtime/src/main/java/gobblin/runtime/Fork.java

```diff
@@ -70,9 +77,27 @@
   private final RowLevelPolicyCheckResults rowLevelPolicyCheckingResult;
   private final DataWriter<Object> writer;
 
+  // This is used to signal the parent task of the completion of this fork
+  private final CountDownLatch countDownLatch;
+
+  // A bounded blocking queue in between the parent task and this fork
+  private final BoundedBlockingRecordQueue<Object> recordQueue;
+
   private final Closer closer = Closer.create();
 
-  public Fork(TaskContext taskContext, TaskState taskState, Object schema, int branches, int index)
+  // This is used by the parent task to signal that it has done pulling records and this fork
+  // should not expect any new incoming data records. This is written by the parent task and
+  // read by this fork. Since this flag will be updated by only a single thread and updates to
+  // a boolean are atomic, volatile is sufficient here.
+  private volatile boolean parentTaskDone = false;
+
+  // This may be updated and read by both this fork and the parent task although the update by
+  // the parent task will only come after the updates by this fork. But to be absolutely safe,
+  // an AtomicReference is still used here to guarantee atomicity.
+  private final AtomicReference<ForkState> forkState;
```

Comment:
According to the Java Language Spec: "Writes to and reads of references are always atomic, regardless of whether they are implemented as 32-bit or 64-bit values." It looks like here we need AtomicReference only for the compareAndSet() operation.